pcm_converter: add implementations

tlauda opened the pull request:
Implements generic and HiFi3 versions of PCM converter library component.
This module provides static functions, which allows to copy
buffer data with frame format conversion.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

tlauda commented:
@singalsu Can you take a look if processing looks OK?

dbaluta commented on src/include/sof/audio/buffer.h:
```diff
@@ -229,15 +229,14 @@ static inline void buffer_init(struct comp_buffer *buffer, uint32_t size,
 
-	for (i = 0; i < frames; i++) {
+	for (i = 0; i < samples; i++) {
 		src = buffer_read_frag_s16(source, buff_frag);
```
What is the difference between frames and samples?

paulstelian97 replied:
@dbaluta Wasn't the frame the collection of samples one per channel? As in, the frame being two samples in case of stereo?

tlauda replied:
@dbaluta @paulstelian97 Exactly, the naming was wrong from the beginning.

tlauda commented:
@xiulipan It builds locally for me. I think it will fail for ICL until we use new toolchain.

dbaluta commented on src/audio/CMakeLists.txt:
```diff
@@ -56,6 +56,7 @@ if(NOT BUILD_LIBRARY)
 			detect_test.c
 		)
```
@tlauda please start the commit message with a brief description of what is "PCM converter" and what is its purpose.

tlauda replied:
@dbaluta But it's there: This module provides static functions, which allows to copy
buffer data with frame format conversion.

dbaluta replied:
You are right.

dbaluta commented on src/audio/pcm_converter/pcm_converter_hifi3.c:
```diff
@@ -0,0 +1,643 @@
+#include <sof/audio/pcm_converter.h>
+
+#ifdef CONFIG_PCM_CONVERTER_HIFI3
+
```
Why do you need this #ifdef here? Just compile out the entire file if CONFIG_HAVE_PCM_CONV_HIFI3 is not set.

tlauda replied:
@dbaluta It's coming from header, not config. Fixed name.

dbaluta commented on src/include/sof/audio/pcm_converter.h:
```diff
@@ -20,6 +20,21 @@
+#include <xtensa/config/core-isa.h>
+
+#if XCHAL_HAVE_HIFI3
+
```
Do, we want to force the usage of the optimized conversion if HIFI3 is set right?

tlauda replied:
I would assume so. It's like that for all the component. In the future we could add option to config to toggle implementations.

jajanusz commented on src/include/sof/audio/pcm_converter.h:
```diff
@@ -0,0 +1,78 @@
+
+struct comp_buffer;
+
+#define CONFIG_PCM_CONVERTER_GENERIC
```
CONFIG_* variables should come only from Kconfig.

singalsu commented on src/audio/pcm_converter/pcm_converter_generic.c:
```diff
@@ -0,0 +1,160 @@
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s16(source, buff_frag);
+		dst = buffer_write_frag_s16(sink, buff_frag);
+		*dst = *src;
```
Have you considered memcpy() based function instead? There's example in src_copy_s16()?

paulstelian97 replied:
@singalsu What about "non-uniform wrapping", as I would call? For example we want to copy 20 samples. After 10 samples the source buffer wraps, after another 5 the destination buffer wraps. I think this code, even if it may be less performant (I'm not very convinced it is, for the record) is simpler to understand.

singalsu replied:
1:1 copy functions could be made available to library to avoid duplication. SRC uses it for same rate source & sink, and EQs use such for bypass mode.

singalsu commented on src/audio/pcm_converter/pcm_converter_generic.c:
```diff
@@ -0,0 +1,160 @@
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s32(source, buff_frag);
+		dst = buffer_write_frag_s16(sink, buff_frag);
+		*dst = *src >> 8;
```
Rounding is missing, also remember to saturate when rounding. Format.h contains sat_int16() function.  Macro Q_SHIFT_RND() could be used for the rounding,

singalsu commented on src/audio/pcm_converter/pcm_converter_generic.c:
```diff
@@ -0,0 +1,160 @@
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s32(source, buff_frag);
+		dst = buffer_write_frag_s32(sink, buff_frag);
+		*dst = *src >> 8;
```
Rounding is missing.

singalsu commented on src/audio/pcm_converter/pcm_converter_hifi3.c:
```diff
@@ -0,0 +1,643 @@
+
+		/* load and shift right one 32 bit samples */
+		AE_L32_XC(sample_1, (ae_int32 *)in, sizeof(ae_int32));
+		sample = AE_MOVINT16X4_FROMINT32X2(AE_SRAI32(sample_1, 8));
```
The AE_SRAI32() right shift is missing the rounding.

singalsu commented on src/audio/pcm_converter/pcm_converter_hifi3.c:
```diff
@@ -0,0 +1,643 @@
+		pcm_converter_setup_circular(sink);
+
+		/* shift right and store four 16 bit samples */
+		sample = AE_CVT16X4(AE_SRAI32(sample_1, 8),
```
Ditto

singalsu commented on src/audio/pcm_converter/pcm_converter_hifi3.c:
```diff
@@ -0,0 +1,643 @@
+		pcm_converter_setup_circular(sink);
+
+		/* store four 16 bit samples */
+		AE_SA16X4_IC(AE_SEL16_6420(sample_1, sample_2), align_out, out);
```
AE_SEL16_xxxx() does not round.

xiulipan commented:
@tlauda PR for the XCC update is already here https://github.com/thesofproject/sof/pull/2024

tlauda commented:
@singalsu I've fixed all rounding problems. Also I've changed basic buffer copy functions to use memcpy_s. My basic tests has shown performance increase by ~20%, but of course it can differ.

lgirdwood commented on src/audio/dai.c:
```diff
@@ -108,16 +108,17 @@ static void dai_dma_cb(void *data, uint32_t type, struct dma_cb_data *next)
-		dma_buffer_copy_to(local_buffer, dd->dma_buffer, dd->process,
-				   bytes);
+		dma_buffer_copy_to(local_buffer, bytes, dd->dma_buffer, bytes,
+				   dd->process, bytes / comp_sample_bytes(dev));
```
division can be expensive, can we pre-calculate this value and just pass samples directly ?

tlauda replied:
@lgirdwood Bytes are different every copy. We need to divide somewhere, so it won't matter whether I move it to processing function or leave it here. It's nothing new, since we divide in almost every processing component.

lgirdwood commented on src/include/sof/audio/pcm_converter.h:
```diff
@@ -0,0 +1,63 @@
+	enum sof_ipc_frame sink;	/**< sink frame format */
+	/**< PCM conversion function */
+	void (*func)(struct comp_buffer *source, struct comp_buffer *sink,
+		     uint32_t frames);
```
frames -> samples ?

tlauda replied:
@lgirdwood Done.

singalsu commented:
> @singalsu I've fixed all rounding problems. Also I've changed basic buffer copy functions to use memcpy_s. My basic tests has shown performance increase by ~20%, but of course it can differ.

That's excellent, thanks @tlauda !

lgirdwood commented:
@fredoh9 checkpatch result is all clear here but CI reporting checpatch error ?

paulstelian97 commented:
> @fredoh9 checkpatch result is all clear here but CI reporting checpatch error ?

I do see the split string error. This PR is from before that change to Checkpatch and I suppose may need updating due to this.

zrombel commented:
Bit depth conversion mechanism is changed and CI test fails. New test are ready and I've checked manually that this PR is passing. From my point of view PR is ready to be merged and new test will be also merged.

lgirdwood commented:
@zrombel thanks for your prompt check. @fredoh9 any update on your side ?

lgirdwood commented:
SOFCI TEST

tlauda commented:
@lgirdwood We need #2024 in order for the ICL compilation to pass.

tlauda commented:
SOFCI TEST

tlauda commented:
@lgirdwood Device test passed. Internal CI will fail anyway as @zrombel reported. Tests will be changed as soon as this PR will be merged.
